internal/storage/roles.go

```diff
@@ -0,0 +1,274 @@
+			resource_id,
+			creator_id,
+			created_at,
+			updated_at
+		FROM roles
+		WHERE
+			resource_id = $1
+		`,
+		resourceID.String(),
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var roles []Role
+
+	for rows.Next() {
+		var role Role
+
+		if err := rows.Scan(&role.ID, &role.Name, &role.ResourceID, &role.CreatorID, &role.CreatedAt, &role.UpdatedAt); err != nil {
+			return nil, err
+		}
+
+		roles = append(roles, role)
+	}
+
+	return roles, nil
+}
+
+// CreateRole creates a role with the provided details.
+// If a role already exists with the given roleID an ErrRoleAlreadyExists error is returned.
+// If a role already exists with the same name under the given resource ID then an ErrRoleNameTaken error is returned.
+//
+// This method must be called with a context returned from BeginContext.
+// CommitContext or RollbackContext must be called afterwards if this method returns no error.
+func (e *engine) CreateRole(ctx context.Context, actorID, roleID gidx.PrefixedID, name string, resourceID gidx.PrefixedID) (Role, error) {
+	tx, err := getContextTx(ctx)
+	if err != nil {
+		return Role{}, err
+	}
+
+	var role Role
+
+	err = tx.QueryRowContext(ctx, `
+		INSERT
+			INTO roles (id, name, resource_id, creator_id, created_at, updated_at)
+			VALUES ($1, $2, $3, $4, now(), now())
+		RETURNING id, name, resource_id, creator_id, created_at, updated_at
+		`, roleID.String(), name, resourceID.String(), actorID.String(),
+	).Scan(
+		&role.ID,
+		&role.Name,
+		&role.ResourceID,
+		&role.CreatorID,
+		&role.CreatedAt,
+		&role.UpdatedAt,
+	)
+
+	if err != nil {
+		if pqIsRoleAlreadyExistsError(err) {
+			return Role{}, fmt.Errorf("%w: %s", ErrRoleAlreadyExists, roleID.String())
+		}
+
+		if pqIsRoleNameTakenError(err) {
+			return Role{}, fmt.Errorf("%w: %s", ErrRoleNameTaken, name)
+		}
+
+		return Role{}, err
+	}
+
+	return role, nil
+}
+
+// UpdateRole updates an existing role if one exists.
+// If no role already exists, a new role is created in the same way as CreateRole.
+// If changing the name and the new name results in a duplicate name error, an ErrRoleNameTaken error is returned.
+//
+// This method must be called with a context returned from BeginContext.
+// CommitContext or RollbackContext must be called afterwards if this method returns no error.
+func (e *engine) UpdateRole(ctx context.Context, roleID gidx.PrefixedID, name string) (Role, error) {
+	tx, err := getContextTx(ctx)
+	if err != nil {
+		return Role{}, err
+	}
+
+	var row *sql.Row
+
+	// If no name is provided, only update the updated_at timestamp.
+	if name == "" {
+		row = tx.QueryRowContext(ctx, `
+			UPDATE roles SET updated_at = now() WHERE id = $1
+			RETURNING id, name, resource_id, creator_id, created_at, updated_at
+		`, roleID.String())
+	} else {
+		row = tx.QueryRowContext(ctx, `
+			UPDATE roles SET name = $1, updated_at = now() WHERE id = $2
+			RETURNING id, name, resource_id, creator_id, created_at, updated_at
+			`, name, roleID.String(),
+		)
+	}
```

Comment:
As before, we can probably make this less complex by just accepting whatever role name we get.